build/gengo.bzl

```diff
@@ -0,0 +1,171 @@
+#             "k8s.io/kubernetes/a/zz_genrated.deepcopy.go/file",
+#           ],
+#           deps=[
+#             "vendor/github.com/every/package",
+#             "vendor/whatever.io/used/by/packages",
+#           ],
+#      )
+#   2) For every package in the packages list, add the following:
+#       load("//build:gengo.bzl", "k8s_gengo")
+#       k8s_gengo(name="generate-deepcopy", outs=["zz_generated.deepcopy.go"])
+#   3) Be sure to use the //build:go.bzl wrapped version of go_library
+#       load("@io_bazel_rules_go//go:def.bzl", "go_library")  # NO: generate-deepcopy fails
+#       load("//build:go.bzl", "go_library")  # YES: generate-deepcopy succeeds
+#   4) gazelle will now notice the existence of this file and add it to srcs
+#       go_library(
+#           name="go_default_library",
+#           srcs=[..., "zz_generated.deepcopy.go"],
+#           ...
+#       )
+
+load("@io_kubernetes_build//defs:go.bzl", _go_genrule="go_genrule")
+
+def _generate_prefix(tool, tool_flags, out, match, header):
+  """Set the variables that change for different generators.
+
+  Args:
+    tool: a gengo tool, something like k8s.io/code-generator/cmd/deepcopy-gen
+    tool_flags: unique flags for the tool, something like --bounding-dirs k8s.io/kubernetes
+    out: the basename of the file, something like zz_generated.deepcopy.go
+    match: search packages for files with this tag, something like +k8s:deepcopy-gen=
+    header: boilerplate file, something like vendor/k8s.io/code-generator/hack/boilerplate.go.txt
+  """
+  return """
+  tool='$(location {tool})'
+  tool_flags='{flags}'
+  out='{out}'
+  match='{match}'
+  head='{header}'
+""".format(tool=tool, tool_flags=tool_flags, out=out, match=match)
+
+def _generate_outs(packages):
+  """Return the move-out commands which copy files to where bazel expects."""
+  return "\n".join(["move-out '%s'" % p for p in packages])
+
+def source_files(pkgs):
+  """Return a list of sourcefiles targets for the specified packages.
+
+  The :go_default_library.sourcefiles target includes all the
+  non-generated *.go files for the package. Both the gengo and go list
+  tools use these files to reflect over the package.
+  """
+  return ["//%s:go_default_library.sourcefiles" % p for p in pkgs]
+
+def k8s_gengo_all(name, base, tool, flags, match, packages, deps):
+  """Use TOOL to generate BASE files in all the PACKAGES with a MATCH comment."""
+  # TODO(fejta): gazelle doesn't appear to auto-generate rules for this package
+  packages = [p for p in packages if not p.startswith("vendor/k8s.io/code-generator/_examples/")]
+  # Add any missing packages to deps
+  deps = {d: True for d in (deps + packages) if not d.startswith("vendor/k8s.io/code-generator/_examples/")}.keys()
+
+  # Tell bazel all the files we will generate
+  go_files = ["%s/%s" % (p, base) for p in packages] # generated files
+  dep_files = ["%s.deps" % g for g in go_files] # list of new packages dependencies
```

Comment:
Looking at your other PR, I think what's going on here is that the generated file adds some go dependencies, so you're trying to express these new dependencies to the `go_library` which depends on this generated file. Am I understanding correctly?

writing these dependencies out during the genrule execution, and then reading those files to rewrite build rules seems pretty smelly. surely there's some better way to find out what dependencies each generated file will need? (maybe I am being too optimistic.)

Reply:
I'm curious: what additional dependencies do these generated files add?

I'm guessing some of them might use some common library or something, but I'd expect that to be common across all generated-code of that type, so we could in theory just add it a dependency on all such rules of that type. Are there examples of additional dependencies being added that aren't common for the gen type and aren't already imported by some other file in that package?

Reply:
I did not want or start with this complexity. I added it after pain, suffering and builds failing to find the right imports.

There are no restrictions on what the generated files import, and they go all over the place:
https://github.com/kubernetes/kubernetes/blob/d65a8ac95cdd635afeb401d04a4fbc496ec58098/pkg/apis/admission/v1beta1/doc.go#L18

This is not that complicated to do and is correct. The alternatives are worse and less correct (like depending on everything, or attaching a bunch of sed commands).

Reply:
your other PR adds dozens of lines of shell and additional technical debt that I worry will get us in a worse situation than we are right now. (it also requires having bazel installed to run `hack/update-bazel.sh`, which is a non-starter, at least for the time being.)

if the additional dependencies are actually as simple as the comment you've listed here, it'd be pretty straightforward for the tool that maintains the `k8s_gengo` rules (i.e. kazel, once implemented) to manage the additional deps needed, too.